### PR TITLE
[CodeGen] Avoid creating instructions dealing with ones if ones aren't involved in the division

### PR DESCRIPTION
**Pull request by AZero13**

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 1]
```diff
@@ -5213,6 +5213,9 @@ MachineInstr *CombinerHelper::buildUDivUsingMul(MachineInstr &MI) {
     return true;
   };
 
+  bool hasAOne = false;
```
Comment (topperc) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:5216]: Variable names should start with capital letter

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 2]
```diff
@@ -6628,11 +6634,14 @@ SDValue TargetLowering::BuildUDIV(SDNode *N, SelectionDAG &DAG,
...
-
-  SDValue One = DAG.getConstant(1, dl, VT);
-  SDValue IsOne = DAG.getSetCC(dl, SetCCVT, N1, One, ISD::SETEQ);
-  return DAG.getSelect(dl, VT, IsOne, N0, Q);
+  if (hasAOne) {
+    EVT SetCCVT =
+        getSetCCResultType(DAG.getDataLayout(), *DAG.getContext(), VT);
+    SDValue One = DAG.getConstant(1, dl, VT);
+    SDValue IsOne = DAG.getSetCC(dl, SetCCVT, N1, One, ISD::SETEQ);
+    return DAG.getSelect(dl, VT, IsOne, N0, Q);
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:6642]: This should all get constant folded inside of getNode so I'm not sure what your goal is.

Comment (AtariDreams) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:6642]: It doesn't always though.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 3]
```diff
@@ -5320,12 +5327,16 @@ MachineInstr *CombinerHelper::buildUDivUsingMul(MachineInstr &MI) {
...
   }
 
-  Q = MIB.buildLShr(Ty, Q, PostShift).getReg(0);
-  auto One = MIB.buildConstant(Ty, 1);
-  auto IsOne = MIB.buildICmp(
-      CmpInst::Predicate::ICMP_EQ,
-      Ty.isScalar() ? LLT::scalar(1) : Ty.changeElementSize(1), RHS, One);
-  return MIB.buildSelect(Ty, IsOne, LHS, Q);
+  if (hasAOne) {
+    Q = MIB.buildLShr(Ty, Q, PostShift).getReg(0);
```
Comment (topperc) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:5331]: If this really removes instructions, you should be able to show that in an MIR test.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp (hunk 2 above)
Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:6642]: Do you have an example?

**Pull request closed without merging**